Changed config for multiple island types, added single block island

LezChap opened the pull request:
* Modifies the config to allow for multiple island types
* Updates old configs to use the new config Properties, preserving settings from old config Properties
* Adds a single Grass Block as an island type
* Removes unneeded import in JoinIslandCommand

(edit to change terminology from "conifg format")

TheCricket commented:
new config format?

LezChap commented:
Instead of a platform=true/false and a skyFactory=true/false, I just made a string config for Platform Type = grass/tree/platform.  It doesn't care about case, but will change it to type platform if the config is misspelled/incorrect.  When it creates a platform, it will create the configured platform type.

As far as I could tell, the platform entry in the config wasn't even used, and it only checked if "Sky Factory" was true.  If it was, it'd make a tree, otherwise it'd create a platform.  I think the new format cleans the config up a bit because of this.  If the config was set up for a tree (even with the unused "platform=false" for tree), it'd change the "Platform Type=tree", otherwise it defaults to changing the config to "Platform Type=platform".  It removes the old two properties (Sky Factory and Platform) and leaves the new "Platform Type" property for the new config format.  It's functionally (for the user) the same way YUNoGoodMap handles the platform type configuration, and would allow for easier adding of future island types that may be requested.

TheCricket commented:
Probably not best to say new config format then and rather do I redesigned your config

LezChap commented:
This new commit requires Botania's jar added to your dev environment to compile, but does not depend on it for FTB-Islands to run on servers or clients.

TheCricket commented:
thats a 1.8 botania

LezChap commented:
That's a 1.7.10 Botania, which for some reason was given a version number of r1.8...I pulled it straight from my Sky Factory instance.  Look at the official Botania website: http://botaniamod.net/downloads.php

TheCricket commented:
i'm tired

LezChap commented:
It's alright, I fully understand the confusion...the version number for Botania is a bit odd.  I also double-checked after your comment (even though I tested and checked before pushing the commit) just to be sure I hadn't made a mistake myself.

The only reason I required Botania at all is it was the only way I could figure out how to get the Botania-unique blocks (Living Root and the Mana Flame) that are part of the GoG island.  I made sure that if Botania isn't installed, it not only doesn't crash, it instead (if GoG is selected as the island type in the config) creates a 3x4x3 island of dirt and grass with a water block on the side (IMO it's close enough to the GoG island type considering it's only using Vanilla blocks). I admit I don't know Java well enough to know if there's a better way that avoids the requirement for Botania being present.

TheCricket commented:
GameRegistry.findItem("modid", "unlocalizedName") btw

LezChap commented:
Well I've played around with the GameRegistry.findBlock method for getting Botania's blocks and not requiring Botania as a build dependency...and everything's working.  My biggest issue right now is I can't figure out a way to create the island without using the same/similar code to how Botania does it without making the code grossly inefficient and illegible, and I'm not sure how you and Kolatra feel about that, and the possible license implications.  

Right now I'm essentially using the same process and code Vazkii is: https://github.com/Vazkii/Botania/blob/master/src/main/java/vazkii/botania/common/world/SkyblockWorldEvents.java#L162, the order blocks are placed is changed a bit, and the variables are shifted on the Y axis by 1, but it's essentially the same, especially with how she uses the int matrix on line 171.  I don't entirely feel comfortable with just changing the variable names, or even the root placement x/y/z's and calling it clean/independent code.

TheCricket commented:
Kolatra has all control over what happens next

Kolatra commented:
I don't see any immediate issues with the PR. It'll require some testing to make sure everything works, I'll see if I can talk to Darkosto as well and give him the chance to test this.

Kolatra commented on src/main/java/com/cricketcraft/ftbisland/IslandCreator.java:
```diff
@@ -62,7 +67,25 @@ public static boolean spawnIslandAt(World world, int x, int y, int z, String pla
+	 		// Creates a Garden of Glass island type
+            } else if (FTBIslands.islandType.equalsIgnoreCase("GoG")) {
+        		if (Loader.isModLoaded("Botania")) {
+        			SkyblockWorldEvents.createSkyblock(world,  x,  y -1,  z);
```
This may want to be changed to the full import, vazkii.botania.common.world.SkyblockWorldEvents.createSkyBlock(world, x, y -1, z); just to avoid import issues if the mod is not loaded.

Kolatra commented:
Actually, Cricket and I think that the Botania soft dependency should be removed, and a similar island can just be created.

LezChap commented:
You'll have a new commit with the dependency on Botania removed shortly.

LezChap commented:
Changed the design of the "roots" a bit from GoG's default (moved a couple, added a couple), made it use oak logs instead of roots if Botania isn't present.  No dependency on Botania to build after the changes in this commit, thanks to TheCricket pointing out the GameRegistry in Forge.

Kolatra commented:
Implemented. Thank you for this. :)
